planner/core/plan_cost.go

```diff
@@ -39,27 +41,130 @@ func (p *basePhysicalPlan) GetPlanCost(taskType property.TaskType) (float64, err
 
 // GetPlanCost calculates the cost of the plan if it has not been calculated yet and returns the cost.
 func (p *PhysicalSelection) GetPlanCost(taskType property.TaskType) (float64, error) {
-	return 0, errors.New("not implemented")
+	if p.planCostInit {
+		return p.planCost, nil
+	}
+	var cpuFactor float64
+	switch taskType {
+	case property.RootTaskType, property.MppTaskType:
+		cpuFactor = p.ctx.GetSessionVars().CPUFactor
+	case property.CopSingleReadTaskType, property.CopDoubleReadTaskType:
+		cpuFactor = p.ctx.GetSessionVars().CopCPUFactor
+	default:
+		return 0, errors.Errorf("unknown task type %v", taskType)
+	}
+	childCost, err := p.children[0].GetPlanCost(taskType)
+	if err != nil {
+		return 0, err
+	}
+	p.planCost = childCost
+	p.planCost += p.children[0].StatsCount() * cpuFactor // selection cost: rows * cpu-factor
+	p.planCostInit = true
+	return p.planCost, nil
 }
 
 // GetPlanCost calculates the cost of the plan if it has not been calculated yet and returns the cost.
 func (p *PhysicalProjection) GetPlanCost(taskType property.TaskType) (float64, error) {
-	return 0, errors.New("not implemented")
+	if p.planCostInit {
+		return p.planCost, nil
+	}
+	childCost, err := p.children[0].GetPlanCost(taskType)
+	if err != nil {
+		return 0, err
+	}
+	p.planCost = childCost
+	p.planCost += p.GetCost(p.StatsCount()) // projection cost
+	p.planCostInit = true
+	return p.planCost, nil
 }
 
 // GetPlanCost calculates the cost of the plan if it has not been calculated yet and returns the cost.
 func (p *PhysicalIndexLookUpReader) GetPlanCost(taskType property.TaskType) (float64, error) {
-	return 0, errors.New("not implemented")
+	if p.planCostInit {
+		return p.planCost, nil
+	}
+	p.planCost = 0
+	// child's cost
+	for _, child := range []PhysicalPlan{p.indexPlan, p.tablePlan} {
+		childCost, err := child.GetPlanCost(taskType)
+		if err != nil {
+			return 0, err
+		}
+		p.planCost += childCost
+	}
+
+	// index-side net I/O cost: rows * row-size * net-factor
+	netFactor := getTableNetFactor(p.tablePlan)
+	rowSize := getTblStats(p.indexPlan).GetAvgRowSize(p.ctx, p.indexPlan.Schema().Columns, true, false)
+	p.planCost += p.indexPlan.StatsCount() * rowSize * netFactor
+
+	// index-side net seek cost
+	p.planCost += estimateNetSeekCost(p.indexPlan)
+
+	// table-side net I/O cost: rows * row-size * net-factor
+	tblRowSize := getTblStats(p.tablePlan).GetAvgRowSize(p.ctx, p.tablePlan.Schema().Columns, false, false)
+	p.planCost += p.tablePlan.StatsCount() * tblRowSize * netFactor
+
+	// consider concurrency
+	p.planCost /= float64(p.ctx.GetSessionVars().DistSQLScanConcurrency())
+
+	// lookup-cost(table-size seek cost)
```

Comment:
Do we miss table-side seek cost here?
https://github.com/pingcap/tidb/blob/5ba2adb5cb9d2d1c795112357433102a57e7d33e/planner/core/task.go#L1076-L1081

Reply:
fixed